Lib/pathlib.py

```diff
@@ -54,13 +54,29 @@ def _ignore_error(exception):
             getattr(exception, 'winerror', None) in _IGNORED_WINERRORS)
 
 
+@functools.lru_cache()
 def _is_case_sensitive(flavour):
     return flavour.normcase('Aa') == 'Aa'
 
 #
 # Globbing helpers
 #
 
+
+# fnmatch.translate() returns a regular expression that includes a prefix and
+# a suffix, which enable matching newlines and ensure the end of the string is
+# matched, respectively. These features are undesirable for our implementation
+# of PurePatch.match(), which represents path separators as newlines and joins
+# pattern segments together. As a workaround, we define a slice object that
+# can remove the prefix and suffix from any translate() result.
```

Comment:
I see the long discussion about this and think you've landed on something acceptable.

My preference would be for the `def _fnmatch_translate_stripping_prefix(pat)` in this file that Alex proposed (maybe `..._without_anchors`), as it expresses the intent more clearly at the point where it is used and keeps the details elsewhere.

But it's only a single use, and the slice is easy enough to find, so I wouldn't worry. Mostly I wanted to think out loud about architecture 😆

Reply:
However, since this is the first comment being encountered, it would be nice to have a bit of an overview of how the new algorithm works. Just the kind of thing you've posted in various issues/PRs, and will likely be in the commit message, but better to have it near the code. I _would_ worry about this one.

Reply:
I've added a "see _compile_pattern_lines()` comment, and then expanded the explanation there. If you wanted, I could move that method directly below the slice definition?